src/cargo/ops/resolve.rs

```diff
@@ -1,30 +1,101 @@
 use std::collections::HashSet;
 
-use core::{PackageId, PackageIdSpec, SourceId, Workspace};
+use core::{PackageId, PackageIdSpec, PackageSet, Source, SourceId, Workspace};
 use core::registry::PackageRegistry;
 use core::resolver::{self, Resolve, Method};
+use sources::PathSource;
 use ops;
-use util::CargoResult;
+use util::{profile, human, CargoResult, ChainError};
 
-/// Resolve all dependencies for the specified `package` using the previous
+/// Resolve all dependencies for the workspace using the previous
 /// lockfile as a guide if present.
 ///
 /// This function will also write the result of resolution as a new
 /// lockfile.
-pub fn resolve_ws(registry: &mut PackageRegistry, ws: &Workspace)
-                   -> CargoResult<Resolve> {
+pub fn resolve_ws<'a>(ws: &Workspace<'a>) -> CargoResult<(PackageSet<'a>, Resolve)> {
+    let mut registry = PackageRegistry::new(ws.config())?;
+    let resolve = resolve_with_registry(ws, &mut registry, true)?;
+    let packages = get_resolved_packages(&resolve, registry);
+    Ok((packages, resolve))
+}
+
+/// Resolves dependencies for some packages of the workspace,
+/// taking into account `paths` overrides and activated features.
+pub fn resolve_ws_precisely<'a>(ws: &Workspace<'a>,
+                                source: Option<Box<Source + 'a>>,
+                                features: &[String],
+                                all_features: bool,
+                                no_default_features: bool,
+                                specs: &[PackageIdSpec])
+                                -> CargoResult<(PackageSet<'a>, Resolve)> {
+    let features = features.iter().flat_map(|s| {
+        s.split_whitespace()
+    }).map(|s| s.to_string()).collect::<Vec<String>>();
+
+    let mut registry = PackageRegistry::new(ws.config())?;
+
+    let mut write_lockfile = true;
+    if let Some(source) = source {
+        if let Some(root_package) = ws.current_opt() {
+            // Avoid writing a lockfile if we are `cargo install`ing a non local package.
+            write_lockfile = false;
+            registry.add_preloaded(root_package.package_id().source_id(), source);
```

Comment:
I'll need some time to wrap my head around these `Source`s, `Registry`s and `RegistrySource`s :) 

Can we perhaps remember this info during workspace creation, instead of threading it as parameters via several function? I am thinking about something along the lines of

```
struct Workspace {
  kind: WorkspaceKind
}

enum WorkspaceKind { 
  Local,
  Temporary(Box<Source>)
}
```